docs/design/2021-03-09-dynamic-privileges.md

```diff
@@ -0,0 +1,282 @@
++               }
++
+                if item.Priv == mysql.AllPriv {
+                        switch stmt.Level.Level {
+                        case ast.GrantLevelGlobal:
+```
+
+This results in the following written to the log file:
+```
+mysql> grant acdc on *.* to u1;
+ERROR 8121 (HY000): privilege check fail
+
+### Attempting to set DYNAMIC privilege: acdc
+```
+
+It might be possible that changes are still required if `ExtendedPriv` is not supported in all contexts (REVOKE, etc).
+
+Note that creating a role with the same name as a DYNAMIC privilege is supported. A `GRANT` statement can be attributed to ROLES when it omits the ON *.* syntax. Thus:
+
+```
+GRANT BINLOG_ADMIN TO u1; // grants the role binlog_admin
+GRANT BINLOG_ADMIN ON *.* TO u1; // grants the dynamic privilege binlog_admin
+```
+
+This same nuance applies to MySQL.
+
+### Documentation Plan
+
+The statement reference pages for each of the affected metadata commands will need to be updated to describe dynamic privileges.
+
+There will also need to be documentation specific to `DYNAMIC` privileges to describe how it works, and the purpose of fine-grained access control.
+
+
+## Test Design
+
+Testing dynamic privileges is a little bit complex because of the various ways privileges can be inherited by users:
+
+* Direct `GRANT` to the user
+* Granting to a role that the user inherits.
+
+### Functional Tests
+
+Unit tests will be added to cover the semantics around role/dynamic privilege precedence, including logical restoring in a different order.
+
+Integration testing needs to test with global kill enabled/disabled.
+
+### Scenario Tests
+
+The use-cases required by the DBaaS team should be validated when combined with `security-enhanced-mode`. They are:
+
+| Account Name | root | cloudAdmin |
+| --------------- | --------------- | --------------- |
+| Backup & Restore to cloud | Y | Y |
+| File privilege | N | N |
+| Read or set variables | Y | Y | 
+| set restricted variables(some of them even can not be read) | N | Y |
+| Read or set restricted system tables | N | Y | 
+| DROP USER cloudAdmin | N | Hard to N(unless some hardcoded) |
+| REVOKE cloudAdmin | N | Hard to N(unless some hardcoded) |
+| Show processlist, Access to threads belong to other users | Y | Y |
+| Change password if the password of SUPER user is forgotten | N | Y |
+| Kill connections belong to cloudAdmin | N | Y |
+| SHUTDOWN / RESTART | N | Y (graceful shutdown on k8s for tidb-server) |
+
+### Compatibility Tests
+
+The introduction of `DYNAMIC` privileges is not expected to introduce any compatibility issues, because backwards compatibility is ensured. However, plugins should migrate to registering their own dynamic privileges and not rely on the use of `SUPER`. This is considered an enhancement, and not included in-scope for the initial introduction of dynamic privileges (which introduces the framework for plugins to use).
```

Comment:
- Should BR, lightning use the new dynamic privilege as well?
- We need to ensure the cluster can be downgraded in case the upgrading is failed due to some reasons, how to downgrade the cluster to an older version?

Reply:
Downgrade should work, but I can explain why/how. The reason is that the code requests for a specific privilege (such as "BACKUP_ADMIN"), which TIDB also allows `SUPER`. In an earlier version of the code the privilege check just requires `SUPER`.

However, if they don't have `SUPER` and explicitly grant `BACKUP ADMIN`, then downgrade will cause a privilege check failure becaues they don't have `SUPER`.

Reply:
The built-in BACKUP command will be changed to BACKUP_ADMIN. Currently lightning is not a server-plugin (IMHO it should be) so it can't register new dynamic privileges, but maybe the recommended credentials required to do a lightning restore could change to be more fine grained.

We should always recommend the minimum set of privileges, so this makes sense to me. But it is not specifically a compatibility problem because backwards compatibility is ensured.